implementations/rust/ockam/ockam_ffi/src/vault.rs

```diff
@@ -423,53 +420,96 @@ pub extern "C" fn ockam_vault_aead_aes_gcm_decrypt(
+                .aead_aes_gcm_decrypt(&ctx, ciphertext_and_tag, &nonce_vec, additional_data)
+                .await?;
+            if plaintext_size < plain.len() as u32 {
+                return Err(FfiError::BufferTooSmall.into());
+            }
+            *plaintext_length = plain.len() as u32;
 
-        unsafe { std::ptr::copy_nonoverlapping(plain.as_ptr(), plaintext, plain.len()) };
-        Ok::<(), Error>(())
-    }) {
-        Ok(_) => FfiOckamError::none(),
-        Err(err) => err.into(),
-    }
+            unsafe { std::ptr::copy_nonoverlapping(plain.as_ptr(), plaintext, plain.len()) };
+            Ok::<(), Error>(())
+        })?;
+        Ok(())
+    })
 }
 
 /// De-initialize an Ockam Vault.
 #[no_mangle]
 pub extern "C" fn ockam_vault_deinit(context: FfiVaultFatPointer) -> FfiOckamError {
-    block_future(async move {
-        match context.vault_type() {
-            FfiVaultType::Software => {
-                let handle = context.handle() as usize;
-                let mut v = SOFTWARE_VAULTS.write().await;
-                if handle < v.len() {
-                    v.remove(handle);
-                    FfiOckamError::none()
-                } else {
-                    FfiError::VaultNotFound.into()
+    handle_panics(|| {
+        block_future(async move {
+            match context.vault_type() {
+                FfiVaultType::Software => {
+                    let handle = context.handle() as usize;
+                    let mut v = SOFTWARE_VAULTS.write().await;
+                    if handle < v.len() {
+                        v.remove(handle);
+                        Ok(())
+                    } else {
+                        Err(FfiError::VaultNotFound)
+                    }
                 }
             }
-        }
+        })?;
+        Ok(())
     })
 }
+
+fn handle_panics<F>(f: F) -> FfiOckamError
+where
+    F: FnOnce() -> Result<(), FfiOckamError>,
+{
+    let result = std::panic::catch_unwind(std::panic::AssertUnwindSafe(f));
+    match result {
+        // No error.
+        Ok(Ok(())) => FfiOckamError::none(),
+        // Failed with a specific ockam error:
+        Ok(Err(e)) => e,
+        // Panicked
+        Err(e) => {
+            // Force an abort if either:
+            //
+            // - `e` panics during its `Drop` impl.
+            // - `FfiOckamError::from(FfiError)` panics.
+            //
+            // Both of these are extremely unlikely, but possible.
+            let panic_guard = AbortOnDrop;
+            drop(e);
+            let ret = FfiOckamError::from(FfiError::UnexpectedPanic);
+            core::mem::forget(panic_guard);
+            ret
+        }
+    }
+}
+
+/// Aborts on drop, used to guard against panics in a section of code.
+///
+/// Correct usage should `mem::forget` this struct after the non-panicking
+/// section.
+struct AbortOnDrop;
+impl Drop for AbortOnDrop {
+    fn drop(&mut self) {
+        eprintln!("Panic from error drop, aborting!");
```

Comment:
In principal, this `eprintln!` could panic (if stderr is closed, for example), but this Drop can only be called in response to a panic, and Rust calls `core::intrinsics::abort` if it runs a `Drop` that panics while it's already panicking.

That is to say, we could implement this `Drop` as `panic!("Panic from error drop, aborting!")`, but this has notably better user experience, as `core::intrinsics::abort` is a SIGILL (on x86), not a SIGABRT.